[WebXR] Rendering is incorrect in on-device HQ capture

djg opened the pull request:
#### 5b6cb5049aac2b44d2b928fdaae5e6625075bf89
<pre>
[WebXR] Rendering is incorrect in on-device HQ capture
<a href="https://bugs.webkit.org/show_bug.cgi?id=274296">https://bugs.webkit.org/show_bug.cgi?id=274296</a>
<a href="https://rdar.apple.com/128248238">rdar://128248238</a>

Reviewed by Mike Wyrzykowski.

When doing on-device HQ capture of WebXR content the rendering is missing the
right eye and the left eye has warping at the edges. This is because
WebXROpaqueFramebuffer was misinterpreting the layered-to-shared layout required
by the Compositor in this mode, where the texture and viewport allocations are
asymmetric.

The core of fixing this was to understand what coordinate space
drawing and blitting operations happen in:
* All drawing to the drawing framebuffer happen in &quot;screen space&quot; of the
  rasterization rate map.
* Blit from shared to layered textures happen in &quot;physical space&quot; of the
  raterization rate map (ie. texture allocation size)

The framebuffer size has been removed from PlatformXR::FrameData::LayerData. The
display framebuffer size is now calculated from the physical sizes instead of
assuming it was twice the width of the framebuffer for layered layout.

Correct viewports in &quot;screen space&quot; are to be passed via
PlatformXR::FrameData::LayerData. Viewport dimensions are no longer converted to
&quot;physical space&quot;.

Additionally:
* Fixed incorrectly using screen size (ie. 4065x5886) in
  -(MTLRasterizationRateLayerDescriptor intWithSampleCount:) to save allocating
  ~180MB of storage for a few hundred floats
* Refactor rasterization rate map descriptor data into
  PlatformXR::FrameData::RateMapDescription to make it clear it&apos;s intended
  use.
* Remove dead code in WebXROpaqueFramebuffer &amp; GraphicsContextGL.

* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
(rx::mtl::RenderCommandEncoder::setViewport): Viewport coordinates are in
rasterization rate map &quot;screen space&quot; and don&apos;t need conversion.
(rx::mtl::RenderCommandEncoder::setScissorRect): Conversion from physical to
screen coordinates requires rounding to the nearest integer, not floor or ceil.
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::drawFramebufferSize const):
(WebCore::WebXROpaqueFramebuffer::displayFramebufferSize const): Deleted.
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
(WebCore::WebXROpaqueFramebuffer::framebuffer const):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::blitSharedToLayered):
(WebCore::WebXROpaqueFramebuffer::drawFramebufferSize const):
(WebCore::WebXROpaqueFramebuffer::drawViewport const):
(WebCore::calcFramebufferSize):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
(WebCore::WebXROpaqueFramebuffer::displayFramebufferSize const): Deleted.
(WebCore::convertViewportToPhysicalCoordinates): Deleted.
(WebCore::WebXROpaqueFramebuffer::calculateViewportShared): Deleted.
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
(WebCore::newRasterizationRateMap):
* Source/WebCore/platform/xr/PlatformXR.h: Remove dead code PlatformXR::Layout
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::frameTimerFired):
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:

Canonical link: <a href="https://commits.webkit.org/278975@main">https://commits.webkit.org/278975@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1fd77679be9a1a182d2a0a652b67c3c66758646d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52155 "Failed to checkout and rebase branch from PR 28695") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31486 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55428 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2869 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54459 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37885 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2576 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/42443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/1840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/54251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/29122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/45011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/23514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/26390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/2285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1037 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/2431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57025 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27280 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/2486 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/28514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/45122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/29426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7622 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28258 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achan00 commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
```diff
@@ -370,31 +349,40 @@ static PlatformXR::Layout displayLayout(const PlatformXR::FrameData::LayerSetupD
     bool framebufferResize = !m_drawAttachments || m_framebufferSize != framebufferSize || m_displayLayout != displayLayout(data);
-    bool foveationChange = !data.horizontalSamples[0].empty() && !data.verticalSamples.empty() && !data.horizontalSamples[1].empty();
-    m_usingFoveation = foveationChange;
+    bool foveationChange = m_usingFoveation ^ data.foveationRateMapDesc.screenSize.isEmpty();
```
Did you mean:
```
bool foveationChange = m_usingFoveation ^ !data.foveationRateMapDesc.screenSize.isEmpty();
```

Maybe this can be clearer as:
```
    bool usingFoveation = !data.foveationRateMapDesc.screenSize.isEmpty();
    bool foveationChange = m_usingFoveation ^ usingFoveation;
```
and later we can do:
```
m_usingFoveation = usingFoveation;
```

achan00 commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
```diff
@@ -403,16 +391,15 @@ bool WebXROpaqueFramebuffer::setupFramebuffer(GraphicsContextGL& gl, const Platf
+        if (!gl.addFoveation(leftPhysicalSize, rightPhysicalSize, frmd.screenSize, frmd.horizontalSamples[0], frmd.verticalSamples, frmd.horizontalSamples[1]))
             return false;
         gl.enableFoveation(m_drawAttachments.colorBuffer);
     }
```
I noticed that we call `disableFoveation()` on the graphics context always in `endFrame()`. Instead of doing it there, should we just update the foveation state here if `foveationChange` is `true`? Something like:

```
    if (foveationChange) {
        if (m_usingFoveation) {
            const auto& frmd = data.foveationRateMapDesc;
            if (!gl.addFoveation(leftPhysicalSize, rightPhysicalSize, frmd.screenSize, frmd.horizontalSamples[0], frmd.verticalSamples, frmd.horizontalSamples[1]))
                return false;
            gl.enableFoveation(m_drawAttachments.colorBuffer);
        } else
            gl.disableFoveation();
    }
```

djg replied:
Foveation enable used to be global state so we wanted to disable it to avoid affecting other GL operations. Mike's changes made the map affect a texture only, so perhaps the enable/disable are redundant? (They aren't expensive)

djg committed:
[WebXR] Rendering is incorrect in on-device HQ capture
https://bugs.webkit.org/show_bug.cgi?id=274296
rdar://128248238

Reviewed by Mike Wyrzykowski.

When doing on-device HQ capture of WebXR content the rendering is missing the
right eye and the left eye has warping at the edges. This is because
WebXROpaqueFramebuffer was misinterpreting the layered-to-shared layout required
by the Compositor in this mode, where the texture and viewport allocations are
asymmetric.

The core of fixing this was to understand what coordinate space
drawing and blitting operations happen in:
* All drawing to the drawing framebuffer happen in "screen space" of the
  rasterization rate map.
* Blit from shared to layered textures happen in "physical space" of the
  raterization rate map (ie. texture allocation size)

The framebuffer size has been removed from PlatformXR::FrameData::LayerData. The
display framebuffer size is now calculated from the physical sizes instead of
assuming it was twice the width of the framebuffer for layered layout.

Correct viewports in "screen space" are to be passed via
PlatformXR::FrameData::LayerData. Viewport dimensions are no longer converted to
"physical space".

Additionally:
* Fixed incorrectly using screen size (ie. 4065x5886) in
  -(MTLRasterizationRateLayerDescriptor intWithSampleCount:) to save allocating
  ~180MB of storage for a few hundred floats
* Refactor rasterization rate map descriptor data into
  PlatformXR::FrameData::RateMapDescription to make it clear it's intended
  use.
* Remove dead code in WebXROpaqueFramebuffer & GraphicsContextGL.

* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
(rx::mtl::RenderCommandEncoder::setViewport): Viewport coordinates are in
rasterization rate map "screen space" and don't need conversion.
(rx::mtl::RenderCommandEncoder::setScissorRect): Conversion from physical to
screen coordinates requires rounding to the nearest integer, not floor or ceil.
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::drawFramebufferSize const):
(WebCore::WebXROpaqueFramebuffer::displayFramebufferSize const): Deleted.
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
(WebCore::WebXROpaqueFramebuffer::framebuffer const):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::blitSharedToLayered):
(WebCore::WebXROpaqueFramebuffer::drawFramebufferSize const):
(WebCore::WebXROpaqueFramebuffer::drawViewport const):
(WebCore::calcFramebufferSize):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
(WebCore::WebXROpaqueFramebuffer::displayFramebufferSize const): Deleted.
(WebCore::convertViewportToPhysicalCoordinates): Deleted.
(WebCore::WebXROpaqueFramebuffer::calculateViewportShared): Deleted.
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
(WebCore::newRasterizationRateMap):
* Source/WebCore/platform/xr/PlatformXR.h: Remove dead code PlatformXR::Layout
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::frameTimerFired):
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:

Canonical link: https://commits.webkit.org/278975@main
